src/common/src/constants.rs

```diff
@@ -36,26 +36,82 @@ pub mod hummock {
 }
 
 pub mod log_store {
-    use crate::types::DataType;
+    use crate::hash::VirtualNode;
+    use crate::row::Row;
+    use crate::types::{DataType, ScalarImpl};
+    use crate::util::sort_util::OrderType;
+
+    pub type SeqIdType = i32;
+    pub type RowOpCodeType = i16;
 
     pub const EPOCH_COLUMN_NAME: &str = "kv_log_store_epoch";
     pub const SEQ_ID_COLUMN_NAME: &str = "kv_log_store_seq_id";
     pub const ROW_OP_COLUMN_NAME: &str = "kv_log_store_row_op";
 
-    /// `epoch`, `seq_id`, `row_op`
-    pub const KV_LOG_STORE_PREDEFINED_COLUMNS: [(&str, DataType); 3] = [
-        (EPOCH_COLUMN_NAME, EPOCH_COLUMN_TYPE),
-        (SEQ_ID_COLUMN_NAME, SEQ_ID_COLUMN_TYPE),
-        (ROW_OP_COLUMN_NAME, ROW_OP_COLUMN_TYPE),
-    ];
-    /// `epoch`, `seq_id`
-    pub const PK_TYPES: [DataType; 2] = [EPOCH_COLUMN_TYPE, SEQ_ID_COLUMN_TYPE];
+    /// Predefined column includes log store pk columns + `row_op`
+    pub const KV_LOG_STORE_PREDEFINED_EXTRA_NON_PK_COLUMNS: [(&str, DataType); 1] =
+        [(ROW_OP_COLUMN_NAME, ROW_OP_COLUMN_TYPE)];
 
     pub const EPOCH_COLUMN_TYPE: DataType = DataType::Int64;
     pub const SEQ_ID_COLUMN_TYPE: DataType = DataType::Int32;
     pub const ROW_OP_COLUMN_TYPE: DataType = DataType::Int16;
 
-    pub const EPOCH_COLUMN_INDEX: usize = 0;
-    pub const SEQ_ID_COLUMN_INDEX: usize = 1;
-    pub const ROW_OP_COLUMN_INDEX: usize = 2;
+    pub trait KvLogStorePk: Clone + Send + Sync + Unpin + 'static {
+        const LEN: usize;
+        const EPOCH_COLUMN_INDEX: usize;
+        const ROW_OP_COLUMN_INDEX: usize;
+        const SEQ_ID_COLUMN_INDEX: usize;
+
+        fn pk_types() -> &'static [DataType];
+        fn pk_names() -> &'static [&'static str];
+        fn pk_ordering() -> &'static [OrderType];
+        fn pk(vnode: VirtualNode, encoded_epoch: i64, seq_id: Option<SeqIdType>) -> impl Row;
```

Comment:
IIUC, the reason why we introduce this trait is to hide the implementation of how PK are defined. However, I think these information are all persisted in table catalog. Would it be simpler to just leverage table catalog? Like other exeuctors, they construct a row and feed it to the state table interface. Inside state table, it will derive pk and dist key based on the table catalog information.

Reply:
The trait is mainly used in `serde.rs`. The `pk_xxx` method is mainly used for sanity check on log store schema. The logic to compute pk cannot be inferred directly from the table catalog, and we should specify this logic in the `pk` method. 

But I suddenly realize that we can have a struct to hold these pk information instead of introducing a new trait and making the type definition complex.
```
struct LogStorePkInfo {
    len: usize,
    types: &'static [DataType],
    ...
    compute_pk: fn(vnode: VirtualNode, encoded_epoch: i64, seq_id: Option<SeqIdType>) -> KvLogStorePkRow
}
```

Reply:
I have changed to use a struct instead of trait to store the pk information